src/decoder/lattice-faster-decoder-combine.h

```diff
@@ -0,0 +1,547 @@
+   types; this is an optimization for speed.
+ */
+template <typename FST, typename Token = decodercombine::StdToken>
+class LatticeFasterDecoderCombineTpl {
+ public:
+  using Arc = typename FST::Arc;
+  using Label = typename Arc::Label;
+  using StateId = typename Arc::StateId;
+  using Weight = typename Arc::Weight;
+  using ForwardLinkT = decodercombine::ForwardLink<Token>;
+
+  using StateIdToTokenMap = typename std::unordered_map<StateId, Token*>;
+  using IterType = typename StateIdToTokenMap::const_iterator;
+
+  // Instantiate this class once for each thing you have to decode.
+  // This version of the constructor does not take ownership of
+  // 'fst'.
+  LatticeFasterDecoderCombineTpl(const FST &fst,
+                                 const LatticeFasterDecoderCombineConfig &config);
+
+  // This version of the constructor takes ownership of the fst, and will delete
+  // it when this object is destroyed.
+  LatticeFasterDecoderCombineTpl(const LatticeFasterDecoderCombineConfig &config,
+                                 FST *fst);
+
+  void SetOptions(const LatticeFasterDecoderCombineConfig &config) {
+    config_ = config;
+  }
+
+  const LatticeFasterDecoderCombineConfig &GetOptions() const {
+    return config_;
+  }
+
+  ~LatticeFasterDecoderCombineTpl();
+
+  /// Decodes until there are no more frames left in the "decodable" object..
+  /// note, this may block waiting for input if the "decodable" object blocks.
+  /// Returns true if any kind of traceback is available (not necessarily from a
+  /// final state).
+  bool Decode(DecodableInterface *decodable);
+
+
+  /// says whether a final-state was active on the last frame.  If it was not, the
+  /// lattice (or traceback) will end with states that are not final-states.
+  bool ReachedFinal() const {
+    return FinalRelativeCost() != std::numeric_limits<BaseFloat>::infinity();
+  }
+
+  /// Outputs an FST corresponding to the single best path through the lattice.
+  /// Returns true if result is nonempty (using the return status is deprecated,
+  /// it will become void).  If "use_final_probs" is true AND we reached the
+  /// final-state of the graph then it will include those as final-probs, else
+  /// it will treat all final-probs as one.  Note: this just calls GetRawLattice()
+  /// and figures out the shortest path.
+  bool GetBestPath(Lattice *ofst,
+                   bool use_final_probs = true);
+
+  /// Outputs an FST corresponding to the raw, state-level
+  /// tracebacks.  Returns true if result is nonempty.
+  /// If "use_final_probs" is true AND we reached the final-state
+  /// of the graph then it will include those as final-probs, else
+  /// it will treat all final-probs as one.
+  /// The raw lattice will be topologically sorted.
+  /// The function can be called during decoding, it will take "next_toks_" map
+  /// and generate the complete token list for the last frame. Then recover it
```

Comment:
I guess you mean:
it will process emitting tokens from "next_toks_" map to generate both emitting and non-emitting tokens to get raw lattices.

Reply:
I think in this function, you process non-emit, after that, you get both emit (you have done before) and non-emit, right?

Reply:
No. The start point is a list of the emitting tokens. And then I generate non-emitting and emitting tokens.

Reply:
No. In GetRawLattice(), we will call ProcessNonemitting() once. The start point is a list of emitting tokens (i.e. "cur_toks_"), and generate non-emitting tokens for "cur_toks_".

Reply:
yes, that's why I wanna make it more explicit and saying: 
it will process non-emitting arcs from "next_toks_" map to get both emission and non-emission tokens for getting raw lattices

Reply:
oh, you said "tokens" previously, so I missed it.